ignition_ros2_control/src/ignition_ros2_control_plugin.cpp

```diff
@@ -0,0 +1,400 @@
+
+namespace ignition_ros2_control
+{
+//////////////////////////////////////////////////
+class IgnitionROS2ControlPluginPrivate
+{
+  /// \brief Entity ID for sensor within Gazebo.
+
+public:
+  ignition::gazebo::Entity entity_;
+
+public:
+  std::shared_ptr<rclcpp::Node> node;
+
+  // Thread where the executor will sping
+  std::thread thread_executor_spin_;
+
+  // Executor to spin the controller
+  rclcpp::executors::MultiThreadedExecutor::SharedPtr executor_;
+
+  // Timing
+  rclcpp::Duration control_period_ = rclcpp::Duration(1, 0);
+
+  // Interface loader
+  boost::shared_ptr<pluginlib::ClassLoader<
+      ignition_ros2_control::IgnitionSystemInterface>> robot_hw_sim_loader_;
+
+  // Controller manager
+  std::shared_ptr<controller_manager::ControllerManager> controller_manager_;
+
+  // String with the robot description
+  // TODO(ahcorde): Add param in plugin tag
+  std::string robot_description_ = "robot_description";
+  // String with the name of the node that contains the robot_description
+  // TODO(ahcorde): Add param in plugin tag
+  std::string robot_description_node_ = "robot_state_publisher";
+
+  // Last time the update method was called
+  rclcpp::Time last_update_sim_time_ros_;
+
+  std::string getURDF(std::string param_name) const
+  {
+    std::string urdf_string;
+
+    using namespace std::chrono_literals;
+    auto parameters_client = std::make_shared<rclcpp::AsyncParametersClient>(
+      node, robot_description_node_);
+    while (!parameters_client->wait_for_service(0.5s)) {
+      if (!rclcpp::ok()) {
+        RCLCPP_ERROR(
+          node->get_logger(), "Interrupted while waiting for %s service. Exiting.",
+          robot_description_node_.c_str());
+        return 0;
+      }
+      RCLCPP_ERROR(
+        node->get_logger(), "%s service not available, waiting again...",
+        robot_description_node_.c_str());
+    }
+
+    RCLCPP_INFO(
+      node->get_logger(), "connected to service!! %s asking for %s",
+      robot_description_node_.c_str(),
+      param_name.c_str());
+
+    // search and wait for robot_description on param server
+    while (urdf_string.empty()) {
+      std::string search_param_name;
+      RCLCPP_DEBUG(node->get_logger(), "param_name %s", param_name.c_str());
+
+      try {
+        auto f = parameters_client->get_parameters({param_name});
+        f.wait();
+        std::vector<rclcpp::Parameter> values = f.get();
+        urdf_string = values[0].as_string();
+      } catch (const std::exception & e) {
+        RCLCPP_ERROR(node->get_logger(), "%s", e.what());
+      }
+
+      if (!urdf_string.empty()) {
+        break;
+      } else {
+        RCLCPP_ERROR(
+          node->get_logger(), "ignition_ros2_control plugin is waiting for model"
+          " URDF in parameter [%s] on the ROS param server.", search_param_name.c_str());
+      }
+      usleep(100000);
+    }
+    RCLCPP_INFO(
+      node->get_logger(), "Recieved urdf from param server, parsing...");
```

Comment:
Nitpick: at this point, we don't know that the URDF is going to be parsed. This is done outside of this function.

```suggestion
      node->get_logger(), "Received URDF from param server.");
```